### ISSUE TITLE
Its very easy to lose your work

### ISSUE THREAD

**Issue by majkinetor** [flameshot-org/flameshot#2055]
I can dedicate 10 minutes adding numbers, text etc, just so that one ESC extra or even ctrl + c can close everything without an option to restore on screen what was last visible.

This is major design mistake. I suggest doing any of bellow things:

1. An option to prevent closing the editor. Add explicit [x] somewhere, maybe next to the ToolSettings.
2. An option to restore last view after you accidentally exited. Right clik on tray and then Reopen last one. Even better, keep a list of several recent so we can recall any of them.

**Comment by borgmanJeremy** [comment]
I have been contemplating option 2 for a while now.

Similar to greenshot where to can save it in a resumable format.

**Comment by veracioux** [comment]
I think it would be great to be able to save and resume progress to/from an SVG file. Or maybe an SVG-based spinoff if we need custom functionality.

**Comment by majkinetor** [comment]
>  I have been contemplating option 2 for a while now.
> I think it would be great to be able to save and resume progress to/from an SVG file

SVG sounds ideal... It doesn't really matter how it is stored locally as long as you can restore it.  

How about we have option 1 ASAP and you can ruminate about option 2 as long as you need. Option 1 will certainly prevent losing work by accident. It happens to me at least once a day.

**Comment by veracioux** [comment]
@majkinetor I can relate. I think we should implement the following strategy as a first resort:
1. If no selection is made and if no edits were made, Esc will close the GUI normally.
2. If the user has made some edits, pressing Esc will show a confirmation dialog
3. The GUI can be closed without confirmation using the close button (which already exists) or using the shortcut that the user has bound to that action (by default `Ctrl+Q`, not easy to hit by accident)

I am against adding a config option, because it is less discoverable and bloats the example config file.

**Comment by majkinetor** [comment]
> If no selection is made and if no edits were made, Esc will close the GUI normally.

Selection is not important, you can quickly recreate it, so I would look just for existence of objects.

> If the user has made some edits, pressing Esc will show a confirmation dialog

As an option, confirm makes most sense. Esc on confirmation should cancel confirmation.  Enter should confirm. ESC+Enter by accident is not likely at all.

> The GUI can be closed without confirmation using the close button (which already exists) or using the shortcut that the user has bound to that action (by default Ctrl+Q, not easy to hit by accident)

I can't see the already existing button. Shortcut should be typical CTRL+W (closes tabs and windows almost everywhere).

> I am against adding a config option, because it is less discoverable and bloats the example config file.

I was thinking about general configuration options such as:

-  [ ] Confirm exit if there are user created objects

With it we don't need GUI closure option, it could always be there and work as you suggested.

**Comment by veracioux** [comment]
@majkinetor

> I can't see the already existing button. Shortcut should be typical CTRL+W (closes tabs and windows almost everywhere).

The button is shown only when a selection is active (looks like an X). The shortcut should work independently of the button, even with no selection. If this is not true for you then it's likely because you've disabled the button and/or shortcut in the config. If you are using version 0.10.1 or earlier, the shortcut may not work when the button is disabled in the config. This has been fixed in the recent development versions.

`Ctrl+Q` is also very typical for a "quit" action. I would be fine with either shortcut, but since `Ctrl+Q` has been a default for a long time, I vote to preserve it and prevent confusing existing users.

> I was thinking about general configuration options such as:

I think this is unnecessary and I vote against it.

Agree with the remaining points.

**Comment by majkinetor** [comment]
> The button is shown only when a selection is active (looks like an X). 

I can't see it on v10.0.1: <details><summary>details</summary>
![image](https://user-images.githubusercontent.com/85767/140951193-748c936f-64c9-4001-9445-1cd02180173b.png)

</details>

> The shortcut should work independently of the button, even with no selection. If this is not true for you then it's likely because you've disabled the button and/or shortcut in the config.

My config <details><summary>details</summary>![image](https://user-images.githubusercontent.com/85767/140952037-fe820a9b-ac1c-41f8-b5bd-b17b3fd8180f.png)
</details>

Can't see anything wrong with it.

**Comment by veracioux** [comment]
@majkinetor These settings are in the "Shortcuts" and "Interface" tabs of the configuration window.

**Comment by majkinetor** [comment]
Yeah, my bad. Thx.

**Comment by mmahmoudian** [comment]
Sorry I'm a bit late to the party.

> > If no selection is made and if no edits were made, Esc will close the GUI normally.
> 
> Selection is not important, you can quickly recreate it, so I would look just for existence of objects.

I agree with this. selection area is not important, but the objects can be.

> I can't see the already existing button.

![image](https://user-images.githubusercontent.com/390889/141118577-77d03c0e-0501-43c7-b3c1-434a8bc852d5.png)


> Shortcut should be typical CTRL+W (closes tabs and windows almost everywhere).

Well, the "quit" is generally bound to <kbd>Ctrl</kbd>+<kbd>q</kbd> (e.g Firefox, Vivaldi, LibreOffice, Kate, Gimp, Inkscape), plus the Ctrl+q is already implemented.


In general I would suggest going with SVG file approach. This would be also useful as we can use the same code and functions to save the screenshot as SVG, plus it is loss-less.

**Comment by szymonk211** [comment]
The same problem is with the "copy the selection into the clipboard" button - it should not turn off the editor without the chance to restore it. The action is called "copy the selection into the clipboard", not "copy the selection and quit". Very misleading.

**Comment by majkinetor** [comment]
Indeed. It happened to me that I misclicked it, especially with dynamic button behavior. 

I prefer that it turns off the editor, as it is obviously the last thing one would do intentionally. With option to return previous editing session I think it shouldn't be any changes (except in the tooltip as @szymonk211 noted).

**Comment by majkinetor** [comment]
I just noticed now that ESC button shortcut is grayed out and can not be disabled:

![image](https://user-images.githubusercontent.com/85767/175312871-b5a5af09-0c0c-4ea6-b1ae-22b63de2831c.png)

This is very weird design choice. Why not making it customizable and/or removable. This issue could be quickly solved that way - one can remove this hotkey and rely on _close_ button to finish editing.

**Comment by rhubarb** [comment]
As a 30-year user of vi/vim and bindings in every IDE, this catches me ALL... THE... TIME...  It's just too easy to hit escape and lose everything.  Option #2 is definitely the ideal : I'd actually like to be able to cancel with Esc, but would like to be able to say 'doh' and recover it.
In the meantime, as a workaround, I'm using bettertouchtool on Macos to capture the escape key and assign it to nothing when Process Name == Flameshot.
I've been using the same trick to prevent the Commit dialog for Git in Intellij closing on my for years now.

**Comment by majkinetor** [comment]
Ah god-damn, I use that trick all the time, and it didn't occur to me to use it here. Thanks @rhubarb .

**Comment by crose-project** [comment]
A 'reopen' of last screenshot, including ability to manipulate the custom objects, would be a huge improvement:

- for those who accidently pressed ESC
- for those (like me) who prefer to copy to clipboard and than to realize that the next application can't import the image via clipboard (the workaround to open an app who can import via clipboard and save from there is pain).

**Comment by mmahmoudian** [comment]
The #3758 was merged and added the confirmation for closing the screenshot upon pressing Esc. This covers the main concern of this issue. I will keep this issue open though because of the interest other devs had on implementing the second half of this feature request [[1](https://github.com/flameshot-org/flameshot/issues/2055#issuecomment-964134389),[2](https://github.com/flameshot-org/flameshot/issues/2055#issuecomment-964172769)]
